src/Microsoft.TemplateEngine.Cli/HelpForTemplateResolution.cs

```diff
@@ -0,0 +1,147 @@
+                        remainingPartialMatches.Add(template.Info.GroupIdentity, templateGroup);
+                    }
+
+                    templateGroup.Add(template);
+                }
+            }
+
+            contextProblemMatchGroups = contextProblemMatches.Values.ToList();
+            remainingPartialMatchGroups = remainingPartialMatches.Values.ToList();
+        }
+
+        private static bool DisplayPartialNameMatchAndContextProblems(string templateName, string context, IReadOnlyList<IReadOnlyList<ITemplateMatchInfo>> contextProblemMatchGroups, IReadOnlyList<IReadOnlyList<ITemplateMatchInfo>> remainingPartialMatchGroups)
+        {
+            bool anythingReported = false;
+            if (contextProblemMatchGroups.Count + remainingPartialMatchGroups.Count > 1)
+            {
+                Reporter.Error.WriteLine(string.Format(LocalizableStrings.AmbiguousInputTemplateName, templateName));
+                anythingReported = true;
+            }
+            else if (contextProblemMatchGroups.Count + remainingPartialMatchGroups.Count == 0)
+            {
+                Reporter.Error.WriteLine(string.Format(LocalizableStrings.NoTemplatesMatchName, templateName));
+                Reporter.Error.WriteLine();
+                return true;
+            }
+
+            // scp (2017-08-28): not certain anything in this loop can happen anymore, due to the way filtering occurs.
+            foreach (IReadOnlyList<ITemplateMatchInfo> templateGroup in contextProblemMatchGroups)
+            {
+                // all templates in a group should have the same context & name
+                if (templateGroup[0].Info.Tags != null && templateGroup[0].Info.Tags.TryGetValue("type", out ICacheTag typeTag))
+                {
+                    MatchInfo? matchInfo = WellKnownSearchFilters.ContextFilter(context)(templateGroup[0].Info);
+                    if ((matchInfo?.Kind ?? MatchKind.Mismatch) == MatchKind.Mismatch)
+                    {
+                        Reporter.Error.WriteLine(string.Format(LocalizableStrings.TemplateNotValidGivenTheSpecifiedFilter, templateGroup[0].Info.Name).Bold().Red());
+                        anythingReported = true;
+                    }
+                }
+                else
+                {   // this really shouldn't ever happen. But better to have a generic error than quietly ignore the partial match.
+                    Reporter.Error.WriteLine(string.Format(LocalizableStrings.GenericPlaceholderTemplateContextError, templateGroup[0].Info.Name).Bold().Red());
+                    anythingReported = true;
+                }
+            }
+
+            if (remainingPartialMatchGroups.Count > 0)
+            {
+                Reporter.Error.WriteLine(LocalizableStrings.TemplateMultiplePartialNameMatches.Bold().Red());
+                anythingReported = true;
+            }
+
+            Reporter.Error.WriteLine();
```

Comment:
It'd be good to only write the empty line if we wrote anything to stderr for cases where consumers look for non-empty stderr to determine failure

Reply:
I guess that we may be guaranteed to write something here anyway, so it might be OK

Reply:
Adding the check, it's possible nothing was written.